pkg/products/rhssouser/reconciler.go

```diff
@@ -243,9 +268,342 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, installation *inte
+		or, err = r.createOrUpdateKeycloakAdmin(user, installation, ctx, serverClient)
+		if err != nil {
+			return integreatlyv1alpha1.PhaseFailed, errors.Wrap(err, "failed to create/update the customer admin user")
+		} else {
+			r.logger.Infof("The operation result for keycloakuser %s was %s", user.UserName, or)
+		}
+	}
+
 	return integreatlyv1alpha1.PhaseCompleted, nil
 }
 
+func identityProviderExists(kcClient keycloakCommon.KeycloakInterface) (bool, error) {
+	provider, err := kcClient.GetIdentityProvider(idpAlias, "master")
+	if err != nil {
+		return false, err
+	}
+	if provider != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// The master realm will be created as part of the Keycloak install. Here we update it to add the openshift idp
+func (r *Reconciler) updateMasterRealm(ctx context.Context, serverClient k8sclient.Client, installation *integreatlyv1alpha1.Installation, host string) (*keycloak.KeycloakRealm, error) {
+	keycloakMasterRealmName := "master"
+
+	kcr := &keycloak.KeycloakRealm{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      keycloakMasterRealmName,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	or, err := controllerutil.CreateOrUpdate(ctx, serverClient, kcr, func() error {
+		kcr.Spec.RealmOverrides = []*keycloak.RedirectorIdentityProviderOverride{
+			{
+				IdentityProvider: idpAlias,
+				ForFlow:          "browser",
+			},
+		}
+
+		kcr.Spec.InstanceSelector = &metav1.LabelSelector{
+			MatchLabels: getInstanceLabels(),
+		}
+
+		kcr.Labels = getMasterLabels()
+
+		kcr.Spec.Realm = &keycloak.KeycloakAPIRealm{
+			ID:          keycloakMasterRealmName,
+			Realm:       keycloakMasterRealmName,
+			Enabled:     true,
+			DisplayName: keycloakMasterRealmName,
+		}
+
+		// The identity providers need to be set up before the realm CR gets
+		// created because the Keycloak operator does not allow updates to
+		// the realms
+		err := r.setupOpenshiftIDP(ctx, installation, kcr, serverClient, "master", host)
+		if err != nil {
+			return errors.Wrap(err, "failed to setup Openshift IDP for user-sso")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create/update keycloak realm: %w", err)
+	}
+	r.logger.Infof("The operation result for keycloakrealm %s was %s", kcr.Name, or)
+
+	return kcr, nil
+}
+
+func (r *Reconciler) createOrUpdateKeycloakAdmin(user keycloak.KeycloakAPIUser, inst *integreatlyv1alpha1.Installation, ctx context.Context, serverClient k8sclient.Client) (controllerutil.OperationResult, error) {
+	selector := &keycloak.KeycloakUser{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("generated-%v", user.UserName),
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	return controllerutil.CreateOrUpdate(ctx, serverClient, selector, func() error {
+		ownerutil.EnsureOwner(selector, inst)
+		selector.Spec.RealmSelector = &metav1.LabelSelector{
+			MatchLabels: getMasterLabels(),
+		}
+		selector.Labels = GetInstanceLabels()
+		selector.Spec.User = user
```

Comment:
We need to add federated identity to ensure the accounts are linked.
We do it here:
https://github.com/integr8ly/integreatly-operator/blob/5db8026f76754e6d73c828cc31f436d4ab0fb961/pkg/products/rhsso/reconciler.go#L751-L757

Reply:
Wait a second, `syncAdminUsersInMasterRealm` sets all the fields for the users, then it gets returned and passed into `createOrUpdateKeycloakAdmin`
I suspect that because of the way how `controllerutil.CreateOrUpdate` works, the fields that are not overridden in CreateOrUpdate func, won't be updated.

Reply:
Ohh, never mind, now I see it - `selector.Spec.User = user`
Please mark this comment thread as resolved. 
Sorry for spam everyone.